File: src/main/java/me/nepnep/nepbot/message/command/commands/info/UrbanDictionary.java
package me.nepnep.nepbot.message.command.commands.info;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import me.nepnep.nepbot.Main;
import me.nepnep.nepbot.message.command.Category;
import me.nepnep.nepbot.message.command.ICommand;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.TextChannel;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
import okhttp3.Request;
import okhttp3.ResponseBody;

import java.io.IOException;
import java.net.URLEncoder;
import java.util.List;

public class UrbanDictionary implements ICommand {

    @Override
    public void execute(List<String> args, GuildMessageReceivedEvent event) {
        TextChannel channel = event.getChannel();
        if (args.size() >= 1) {
            String api = "http://api.urbandictionary.com/v0/define?term=";
            String query = String.join(" ", args.subList(0, args.size()));
            ResponseBody body = null;
            try {
                String safeQuery = URLEncoder.encode(query, "UTF-8");
                Request request = new Request.Builder().url(api + safeQuery).build();
                body = Main.HTTP_CLIENT.newCall(request).execute().body();
                ObjectMapper mapper = new ObjectMapper();

                JsonNode json = mapper.readTree(body.string());
                JsonNode list = json.get("list");

                if (list == null) {
                    body.close();
                    channel.sendMessage("Not found or unknown error").queue();
                    return;
                }

                JsonNode best = list.get(0);

                if (best == null) {
                    body.close();
                    channel.sendMessage("Not found").queue();
                    return;
                }

                String definition = best.get("definition").asText();
                String example = best.get("example").asText();

                String safeDefinition = definition.replaceAll("nigg|fag", "[Removed]");
                String safeExample = example.replaceAll("nigg|fag", "[Removed]");

                EmbedBuilder builder = new EmbedBuilder();
                builder.addField("Definition:", safeDefinition, false);
                builder.addField("Example:", safeExample, false);

                channel.sendMessage(builder.build()).queue();
            } catch (IOException e) {
                channel.sendMessage("API Issues, try again later.").queue();
            } catch (IllegalArgumentException e) {
                channel.sendMessage("Embed is too large!").queue();
            } finally {
                if (body != null) {
                    body.close();
                }
            }
        } else {
            channel.sendMessage("Invalid usage!").queue();
        }
    }

    @Override
    public String getInvoke() {
        return "urban";
    }

    @Override
    public Permission requiredPermission() {
        return null;
    }

    @Override
    public Category getCategory() {
        return Category.INFO;
    }

    @Override
    public String getDescription() {
        return "Get a thing's definition from urban dictionary! ;urban <query>";
    }
}

File: src/main/kotlin/me/nepnep/nepbot/message/command/commands/info/UrbanDictionary.kt
package me.nepnep.nepbot.message.command.commands.info

import com.fasterxml.jackson.databind.ObjectMapper
import dev.minn.jda.ktx.messages.Embed
import me.nepnep.nepbot.message.command.AbstractCommand
import me.nepnep.nepbot.message.command.Category
import me.nepnep.nepbot.request
import net.dv8tion.jda.api.entities.channel.middleman.GuildMessageChannel
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import java.net.URLEncoder

class UrbanDictionary : AbstractCommand(
    "urban",
    Category.INFO,
    "Get a thing's definition from urban dictionary! ;urban <query>"
) {
    override suspend fun execute(args: List<String>, event: MessageReceivedEvent, channel: GuildMessageChannel) {
        if (args.isEmpty()) {
            channel.sendMessage("Invalid usage").queue()
            return
        }
        val api = "https://api.urbandictionary.com/v0/define?term="
        val query = URLEncoder.encode(args.subList(0, args.size).joinToString(" "), "UTF-8") // Wtf kotlin

        event.jda.httpClient.request(api + query, {
            val json = ObjectMapper().readTree(it.body!!.string())

            val list = json["list"]
            if (list == null) {
                channel.sendMessage("Not found or unknown error").queue()
                return@request
            }

            val best = list[0]
            if (best == null) {
                channel.sendMessage("Not found").queue()
                return@request
            }

            val definition = best["definition"].textValue()
            val example = best["example"].textValue()

            if (definition.length > MessageEmbed.VALUE_MAX_LENGTH || example.length > MessageEmbed.VALUE_MAX_LENGTH) {
                channel.sendMessage("Embed field is too large!").queue()
                return@request
            }

            val embed = Embed { 
                field("Definition:", definition, false)
                field("Example:", example, false)
            }

            if (embed.isSendable) {
                channel.sendMessageEmbeds(embed).queue()
                return@request
            }
            channel.sendMessage("Embed is too large!").queue()
        }) {
            channel.sendMessage("API Issues, try again later.").queue()
        }
    }
}